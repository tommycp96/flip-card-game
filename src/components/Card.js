import React, { useRef } from 'react';
import { StyleSheet, Text, TouchableOpacity, Animated } from 'react-native';
import { colors } from '../styles/common';

const Card = ({ char, onPress, index, isMatched, isShow }) => {
  const flipAnimation = useRef(new Animated.Value(0)).current;
  let flipRotation = 0;
  flipAnimation.addListener(({ value }) => (flipRotation = value));

  const flipToFrontStyle = {
    transform: [
      {
        rotateY: flipAnimation.interpolate({
          inputRange: [0, 180],
          outputRange: ['0deg', '180deg']
        })
      }
    ]
  };

  const flipToBackStyle = {
    transform: [
      {
        rotateY: flipAnimation.interpolate({
          inputRange: [0, 180],
          outputRange: ['180deg', '360deg']
        })
      }
    ]
  };

  const flipToFront = () => {
    Animated.timing(flipAnimation, {
      toValue: 180,
      duration: 300,
      useNativeDriver: true
    }).start();
  };

  const flipToBack = () => {
    Animated.timing(flipAnimation, {
      toValue: 0,
      duration: 300,
      useNativeDriver: true
    }).start();
  };

  const flipCard = () => {
    if (!isMatched) {
      flipAnimation.setValue(0);
      !!flipRotation ? flipToBack() : flipToFront();
    }
  };

  const handlePress = () => {
    if (isMatched) return;
    onPress(index);
    flipCard();
  };

  const cardTextValue = () => {
    if (isMatched) return <Text style={styles.matched}>{char}</Text>;
    else if (isShow) return <Text style={styles.shown}>{char}</Text>;
    else return <Text style={styles.hidden}>?</Text>;
  };

  return (
    <TouchableOpacity style={styles.container} onPress={handlePress}>
      <Animated.View
        style={{
          ...styles.card,
          ...flipToFrontStyle,
          backgroundColor:
            !isMatched && !isShow ? colors.primary : colors.secondary
        }}
      >
        {cardTextValue()}
      </Animated.View>
      <Animated.View
        style={{
          ...styles.card,
          ...styles.cardBack,
          ...flipToBackStyle,
          backgroundColor:
            !isMatched && !isShow ? colors.primary : colors.secondary
        }}
      >
        {cardTextValue()}
      </Animated.View>
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  container: {
    width: '24%',
    minHeight: 100,
    alignItems: 'center',
    justifyContent: 'center'
  },
  card: {
    width: '100%',
    padding: 8,
    marginBottom: 8,
    borderRadius: 8,
    justifyContent: 'center',
    alignItems: 'center',
    minHeight: 100,
    shadowColor: '#000',
    shadowOffset: {
      width: 2,
      height: 4
    },
    shadowOpacity: 0.22,
    shadowRadius: 2.22,
    elevation: 8,
    backfaceVisibility: 'hidden'
  },
  cardBack: {
    position: 'absolute',
    top: 0
  },
  matched: {
    color: 'lightgrey',
    fontSize: 20
  },
  shown: {
    fontWeight: 'bold',
    fontSize: 20
  },
  hidden: {
    fontSize: 20
  }
});

export default Card;
