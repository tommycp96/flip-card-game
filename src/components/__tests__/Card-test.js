import 'react-native';
import React from 'react';
import Card from '../Card';
import renderer from 'react-test-renderer';

it('should renders correctly', () => {
  const tree = renderer.create(<Card />).toJSON();
  expect(tree).toMatchSnapshot();
});
