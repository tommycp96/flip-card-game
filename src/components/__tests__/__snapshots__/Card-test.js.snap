// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should renders correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "justifyContent": "center",
      "minHeight": 100,
      "opacity": 1,
      "width": "24%",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backfaceVisibility": "hidden",
        "backgroundColor": "#4fc3f7",
        "borderRadius": 8,
        "elevation": 4,
        "justifyContent": "center",
        "marginBottom": 8,
        "minHeight": 100,
        "padding": 8,
        "shadowColor": "#000",
        "shadowOffset": Object {
          "height": 4,
          "width": 0,
        },
        "shadowOpacity": 0.22,
        "shadowRadius": 2.22,
        "transform": Array [
          Object {
            "rotateY": "0deg",
          },
        ],
        "width": "100%",
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 20,
        }
      }
    >
      ?
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "backfaceVisibility": "hidden",
        "backgroundColor": "#4fc3f7",
        "borderRadius": 8,
        "elevation": 4,
        "justifyContent": "center",
        "marginBottom": 8,
        "minHeight": 100,
        "padding": 8,
        "position": "absolute",
        "shadowColor": "#000",
        "shadowOffset": Object {
          "height": 4,
          "width": 0,
        },
        "shadowOpacity": 0.22,
        "shadowRadius": 2.22,
        "top": 0,
        "transform": Array [
          Object {
            "rotateY": "180deg",
          },
        ],
        "width": "100%",
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 20,
        }
      }
    >
      ?
    </Text>
  </View>
</View>
`;
